src/stability_sdk/animation.py

```diff
@@ -670,3 +748,105 @@ def transform_video(self, frame_idx) -> Optional[np.ndarray]:
+        results = self.api.generate(
+            prompts, weights, 
+            args.width, args.height, 
+            steps = adjusted_steps,
+            seed = args.seed,
+            cfg_scale = args.cfg_scale,
+            sampler = sampler, 
+            init_image = init, 
+            init_strength = strength,
+            init_noise_scale = self.frame_args.noise_scale_series[frame_idx], 
+            mask = mask if mask is not None else self.mask,
+            masked_area_init = generation.MASKED_AREA_INIT_ORIGINAL,
+            mask_fixup = True,
+            guidance_preset = guidance,
+        )
+        return results[generation.ARTIFACT_IMAGE][0]
+
+    def _span_render(self, start: int, end: int, prev_frame: Optional[np.ndarray]) -> Generator[Tuple[int, np.ndarray], None, None]:
+        args = self.args
+
+        def apply_xform(frame: np.ndarray, xform: matrix.Matrix) -> Tuple[np.ndarray, np.ndarray]:
+            if args.animation_mode == '2D':
+                xform = to_3x3(xform)
+            frames, masks = self.api.transform([frame], resample_op(args.border, xform, export_mask=True))
+            return frames[0], masks[0]
+
+        if prev_frame is None:
+            prev_frame = self._span_render_frame(start, None)
+
+        # transform the previous frame forward
+        accum_xform = matrix.identity
+        forward_frames, forward_masks = [], []
+        for frame_idx in range(start, end):
+            accum_xform = matrix.multiply(self.build_frame_xform(frame_idx), accum_xform)
+            frame, mask = apply_xform(prev_frame, accum_xform)
+            forward_frames.append(frame)
+            forward_masks.append(mask)
+
+        # inpaint the final frame
+        if not np.all(forward_masks[-1]):
+            forward_frames[-1] = self.inpaint_frame(end-1, forward_frames[-1], forward_masks[-1])
+
+        # run diffusion on top of the final result to allow content to evolve over time
+        strength = max(0.0, self.frame_args.strength_series[end-1])
+        if strength < 1.0:
+            final_frame = self._span_render_frame(end-1, forward_frames[-1])
+        else:
+            final_frame = forward_frames[-1]
+
+        # go backwards through the frames in the span        
+        backward_frames, backward_masks = [final_frame], [np.full_like(forward_masks[-1], 255)]
+        accum_xform = matrix.identity
+        for frame_idx in range(end-2, start-1, -1):
+            frame_xform = self.build_frame_xform(frame_idx+1)
+            inv_xform = np.linalg.inv(frame_xform).tolist()
+            accum_xform = matrix.multiply(inv_xform, accum_xform)
+            xformed, mask = apply_xform(backward_frames[-1], accum_xform)
+            backward_frames.insert(0, xformed)
+            backward_masks.insert(0, mask)
+
+        # inpaint the backwards frame
+        if not np.all(backward_masks[0]):
+            backward_frames[0] = self.inpaint_frame(start, backward_frames[0], backward_masks[0])
```

Comment:
I probably have a misunderstanding about the algorithm. Could you please explain if I'm wrong. Let's consider a case of camera translation in the span [`start`, `end`] for x axis to be {10, 10, 10, -10, -10, -10}, `diffusion_cadence=6`. For both forward and backward passes `accum_xform` will be equal to an identity transform, meaning that the inpaint function will not be called, and the rightmost areas for `frame_idx=start+3` will be warped, not inpainted.
Or my logic is spoiled?

Reply:
Yes you are correct with that camera motion that returns to where it started within the span.